From davej  Fri Apr 11 20:06:57 2008
Return-Path: <fedora-kernel-list-bounces@redhat.com>
X-Spam-Checker-Version: SpamAssassin 3.2.4 (2008-01-01) on
	quoit.install.boston.redhat.com
X-Spam-Level: 
X-Spam-Status: No, score=-2.5 required=5.0 tests=AWL,BAYES_00,
	UNPARSEABLE_RELAY autolearn=ham version=3.2.4
Received: from pobox.devel.redhat.com [10.11.255.8]
	by quoit.install.boston.redhat.com with IMAP (fetchmail-6.3.8)
	for <davej@localhost> (single-drop); Fri, 11 Apr 2008 20:06:57 -0400 (EDT)
Received: from pobox.devel.redhat.com ([unix socket])
	 by pobox.devel.redhat.com (Cyrus v2.2.12-Invoca-RPM-2.2.12-8.1.RHEL4) with LMTPA;
	 Fri, 11 Apr 2008 20:03:31 -0400
X-Sieve: CMU Sieve 2.2
Received: from int-mx1.corp.redhat.com (int-mx1.corp.redhat.com [172.16.52.254])
	by pobox.devel.redhat.com (8.13.1/8.13.1) with ESMTP id m3C03VUw021673;
	Fri, 11 Apr 2008 20:03:31 -0400
Received: from mx1.util.phx.redhat.com (mx1.util.phx.redhat.com [10.8.4.92])
	by int-mx1.corp.redhat.com (8.13.1/8.13.1) with ESMTP id m3C03TPo021615;
	Fri, 11 Apr 2008 20:03:30 -0400
Received: from hormel.redhat.com (hormel1.util.phx.redhat.com [10.8.4.116])
	by mx1.util.phx.redhat.com (8.13.1/8.13.1) with ESMTP id m3C03Oxg016566;
	Fri, 11 Apr 2008 20:03:24 -0400
Received: from listman.util.phx.redhat.com (listman.util.phx.redhat.com [10.8.4.110])
	by hormel.redhat.com (Postfix) with ESMTP id 6D8876188FB;
	Fri, 11 Apr 2008 20:03:24 -0400 (EDT)
Received: from int-mx1.corp.redhat.com (int-mx1.corp.redhat.com
	[172.16.52.254])
	by listman.util.phx.redhat.com (8.13.1/8.13.1) with ESMTP id
	m3C03M0S023637 for <fedora-kernel-list@listman.util.phx.redhat.com>;
	Fri, 11 Apr 2008 20:03:22 -0400
Received: from file.rdu.redhat.com (file.rdu.redhat.com [10.11.255.147])
	by int-mx1.corp.redhat.com (8.13.1/8.13.1) with ESMTP id m3C03Mad021593;
	Fri, 11 Apr 2008 20:03:22 -0400
Received: from [10.13.248.64] (vpn-248-64.boston.redhat.com [10.13.248.64])
	by file.rdu.redhat.com (8.13.1/8.13.1) with ESMTP id m3C03KPG005232;
	Fri, 11 Apr 2008 20:03:21 -0400
Message-ID: <47FFFC3A.2040708@RedHat.com>
Date: Fri, 11 Apr 2008 20:03:06 -0400
From: Steve Dickson <SteveD@redhat.com>
User-Agent: Thunderbird 2.0.0.12 (X11/20080226)
MIME-Version: 1.0
To: Linux NFS Mailing list <linux-nfs@vger.kernel.org>
Content-Type: text/plain; charset=ISO-8859-1
Content-Transfer-Encoding: 7bit
X-Scanned-By: MIMEDefang 2.58 on 172.16.52.254
X-Scanned-By: MIMEDefang 2.58 on 172.16.52.254
X-loop: fedora-kernel-list@redhat.com
Cc: Fedora Kernel Mailing list <fedora-kernel-list@redhat.com>
Subject: [PATCH] NFS Client mounts hang when exported directory do not exist
X-BeenThere: fedora-kernel-list@redhat.com
X-Mailman-Version: 2.1.5
Precedence: junk
List-Id: "Fedora kernel development." <fedora-kernel-list.redhat.com>
List-Unsubscribe: <https://www.redhat.com/mailman/listinfo/fedora-kernel-list>, 
	<mailto:fedora-kernel-list-request@redhat.com?subject=unsubscribe>
List-Archive: <https://www.redhat.com/archives/fedora-kernel-list>
List-Post: <mailto:fedora-kernel-list@redhat.com>
List-Help: <mailto:fedora-kernel-list-request@redhat.com?subject=help>
List-Subscribe: <https://www.redhat.com/mailman/listinfo/fedora-kernel-list>, 
	<mailto:fedora-kernel-list-request@redhat.com?subject=subscribe>
Sender: fedora-kernel-list-bounces@redhat.com
Errors-To: fedora-kernel-list-bounces@redhat.com
Status: RO
Content-Length: 1546
Lines: 43

This patch fixes a regression that was introduced by the string based mounts. 

nfs_mount() statically returns -EACCES for every error returned
by the remote mounted. This is incorrect because -EACCES is
an non-fatal error to the mount.nfs command. This error causes
mount.nfs to retry the mount even in the case when the exported
directory does not exist. 

This patch maps the errors returned by the remote mountd into
valid errno values, exactly how it was done pre-string based 
mounts. By returning the correct errno enables mount.nfs 
to do the right thing.  

Signed-off-by: Steve Dickson <steved@redhat.com>
---
Take 2- Why reinvent the wheel, as Trond pointed out using
nfs_stat_to_errno() makes more sense and is makes things much 
similar, something I'm always a fan of... 

Index: 2008-devel/fs/nfs/mount_clnt.c
===================================================================
--- 2008-devel.orig/fs/nfs/mount_clnt.c
+++ 2008-devel/fs/nfs/mount_clnt.c
@@ -14,6 +14,7 @@
 #include <linux/sunrpc/clnt.h>
 #include <linux/sunrpc/sched.h>
 #include <linux/nfs_fs.h>
+#include "internal.h"
 
 #ifdef RPC_DEBUG
 # define NFSDBG_FACILITY	NFSDBG_MOUNT
@@ -98,7 +99,7 @@ out_call_err:
 
 out_mnt_err:
 	dprintk("NFS: MNT server returned result %d\n", result.status);
-	status = -EACCES;
+	status = -nfs_stat_to_errno(result.status);
 	goto out;
 }
 
